dquartic/dia-nn to massdash feature extraction/Feature_extraction.py

```diff
@@ -0,0 +1,397 @@
+        self.extraction_config = TargetedDIAConfig()
+        self.extraction_config.rt_window = self.rt_window
+        self.extraction_config.mz_tol = self.mz_tol
+        logging.info(f"Extraction configured with RT window: {self.rt_window}s, m/z tolerance: {self.mz_tol} ppm")
+
+    def run_batch_processing(self):
+        """
+        Process peptides in batches to extract MS1 and MS2 features using GPU acceleration.
+
+        Steps:
+            1. Load unique peptide-charge pairs from the report.
+            2. Split the peptides into batches based on the configured batch size.
+            3. For each peptide, extract features via the MassDASH loader.
+            4. Convert the feature data to a GPU DataFrame (cuDF) for accelerated filtering.
+            5. Extract and rename MS1 (precursor) and MS2 (fragment) features.
+            6. Save each batch as CSV files and combine all batches into final output files.
+
+        Returns:
+            dict: Summary of processing including total peptides and output file paths.
+        """
+        logging.info(f"Loading peptides from {self.report_path}")
+        report_df = pd.read_parquet(self.report_path)
+        unique_peptides = report_df[['Modified.Sequence', 'Precursor.Charge']].drop_duplicates()
+        logging.info(f"Total unique peptides: {len(unique_peptides)}")
+        
+        # Split unique peptides into batches.
+        peptide_batches = np.array_split(unique_peptides, int(np.ceil(len(unique_peptides) / self.batch_size)))
+        logging.info(f"Split peptides into {len(peptide_batches)} batches with batch size {self.batch_size}")
+        
+        ms1_all_batches = []
+        ms2_all_batches = []
+        
+        # Process each batch sequentially
+        for batch_idx, batch in enumerate(peptide_batches):
+            logging.info(f"\nProcessing batch {batch_idx+1}/{len(peptide_batches)} with {len(batch)} peptides")
+            batch_start_time = time.time()
+            ms1_list = []
+            ms2_list = []
+            
+            for idx, row in batch.iterrows():
+                peptide = row['Modified.Sequence']
+                charge = int(row['Precursor.Charge'])
+                logging.info(f"  Processing peptide {peptide} (charge {charge}, index in batch: {idx})...")
+                
+                try:
+                    start_extraction = time.time()
+                    feature_map_collection = self.loader.loadFeatureMaps(peptide, charge, self.extraction_config)
+                    elapsed_extraction = time.time() - start_extraction
+                    logging.info(f"    Extraction done in {elapsed_extraction:.2f} seconds.")
+                except Exception as e:
+                    logging.error(f"    Error loading {peptide} (charge {charge}): {e}")
+                    continue
+                
+                # Process each run's feature map
+                for run_name, feature_map in feature_map_collection.items():
```

Comment:
Is there anyway to vectorize this/ remove the 3 level nested for loop? We also don't seem to be doing any parallelization. Could we parallelize the 3rd level for loop to process runs in parallel for the current peptide_charge?